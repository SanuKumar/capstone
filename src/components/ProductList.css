.card-title {
  margin-bottom: -0.5rem !important;
}

input {
  margin-right: 1rem;
}

/* .row>*:last-child {
  background: #ff0000;
} */
