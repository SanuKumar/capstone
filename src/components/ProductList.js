import React, { useState } from 'react'
import { Container, Card, Button, Row, Col, Modal } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import Loader from './Loader';
import "./ProductList.css"
import ReactStars from "react-rating-stars-component";
import { ToastContainer, toast } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css'
import axios from "axios"
import { RiDeleteBin6Line } from 'react-icons/ri';
import { AiOutlineEdit } from 'react-icons/ai';
import { CgAddR } from 'react-icons/cg';

const ProductList = ({ products, loading, isUserLoggedIn, fetchProductCallBack }) => {
  let history = useHistory();
  const [delMulProduct, setDelMulProduct] = useState([])
  const [show, setShow] = useState(false);
  const handleClose = () => setShow(false);
  const [selectedId, setSelectedId] = useState("")
  const [customField, setCustomField] = useState({
    name: true,
    manufacture: true,
    rating: true,
    description: true
  })

  const handleAddProduct = () => {
    if (isUserLoggedIn) {
      history.push('/addproduct')
    } else {
      toast.error("Please login to add product", {
        position: toast.POSITION.TOP_CENTER
      })
    }
  }

  const handleEditProduct = (id) => {
    if (isUserLoggedIn) {
      history.push({
        pathname: `/product/${id}`,
        state: { edit: true }
      })
    }
    else {
      toast.error("Please login to edit product", {
        position: toast.POSITION.TOP_CENTER
      });
    }
  }

  const timeout = (time) => {
    return new Promise(res => setTimeout(res, time));
  }


  const handleViewProduct = (id) => {
    if (isUserLoggedIn) {
      history.push(`/product/${id}`)
    } else {
      toast.error("Please login to view product details", {
        position: toast.POSITION.TOP_CENTER
      })
    }
  }

  const onSiteChanged = (e) => {
    const { name, checked } = e.target
    setCustomField({
      ...customField,
      [name]: checked ? true : false
    })
  }
  const handleShow = (id) => {
    if (isUserLoggedIn) {
      setShow(true)
      setSelectedId(id)
    } else {
      toast.error("Please login to delete product", {
        position: toast.POSITION.TOP_CENTER
      })
    }
  };

  const handleDeleteProduct = async () => {
    setShow(false)
    try {
      let res = await axios.delete(`http://localhost:3002/products/${selectedId}`)
      if (res.status === 200) {
        toast.success("Product deleted successfully!!", {
          position: toast.POSITION.TOP_CENTER
        });
        await timeout(500);
        fetchProductCallBack()
      }
    } catch (error) {
      toast.error("Error while deleting product", {
        position: toast.POSITION.TOP_CENTER
      });
    }
  }

  const selectMultipleProduct = (id) => {
    if (delMulProduct.includes(id)) {
      let remove = delMulProduct.filter((p) => p != id)
      return setDelMulProduct(remove)
    }
    setDelMulProduct([...delMulProduct, id])
  }

  const deleteMultipleProducts = async () => {
    if (isUserLoggedIn) {
      if (window.confirm('Are you sure want to delete selected products')) {
        try {
          delMulProduct.map((did) => {
            axios.delete(`http://localhost:3002/products/${did}`)
          })
          toast.success("Product's deleted successfully!!", {
            position: toast.POSITION.TOP_CENTER
          });
          await timeout(500);
          fetchProductCallBack()
          setDelMulProduct([])
        } catch (error) {
          toast.error("Error while deleting product", {
            position: toast.POSITION.TOP_CENTER
          });
        }
      }
    } else {
      toast.error("Please login to delete products", {
        position: toast.POSITION.TOP_CENTER
      })
    }
  }


  return (
    <>
      <ToastContainer autoClose={1000} />
      <Container className='container-wrapper'>
        <Row>
          <Col lg={10} md={10} sm={12}>
            <h2>Customise Product Fields</h2>
            <form className='form-wrapper'>
              <label>Product Name</label>
              <input
                type="checkbox"
                name="name"
                checked={customField.name}
                onChange={onSiteChanged}
              />
              <label>Manufacture</label>
              <input
                type="checkbox"
                name="manufacture"
                checked={customField.manufacture}
                onChange={onSiteChanged}
              />
              <label>Rating</label>
              <input
                type="checkbox"
                name="rating"
                checked={customField.rating}
                onChange={onSiteChanged}
              />
              <label>Description</label>
              <input
                type="checkbox"
                name="description"
                checked={customField.description}
                onChange={onSiteChanged}
              />
            </form>
          </Col>
          <Col>
            <Row>
              <Col className='m-1'>
                <div style={{ margin: "2px", width: "10.5rem" }}>
                  {delMulProduct.length > 1 &&
                    <Button variant="danger" onClick={deleteMultipleProducts}>Delete Product's <RiDeleteBin6Line style={{ marginTop: "-2px" }} /></Button>
                  }
                </div>
              </Col>
              <Col className='m-1'>
                <div style={{ margin: "2px", width: "9rem" }}>
                  <Button onClick={handleAddProduct}>Add Product <CgAddR style={{ marginTop: "-2px" }} /></Button>
                </div>
              </Col>
            </Row>
          </Col>
        </Row>
      </Container>
      <Modal show={show} onHide={handleClose} centered>
        <Modal.Header closeButton>
          <Modal.Title>Delete Product</Modal.Title>
        </Modal.Header>
        <Modal.Body>Woohoo, Are you sure want to delete the Product!</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Close
          </Button>
          <Button variant="primary" onClick={handleDeleteProduct}>
            Delete
          </Button>
        </Modal.Footer>
      </Modal>
      <br />
      {
        loading ?
          <Loader /> :
          <Container>
            <Row className="justify-content-md-center">
              {products ? products.map((p) => {
                return (
                  <Col xs={12} sm={6} md={6} lg={3} key={p.id}>
                    <Card key={p.id} className="card-wrapper">
                      <Card.Img
                        variant="top"
                        onError={({ currentTarget }) => {
                          currentTarget.onerror = null; // prevents looping
                          currentTarget.src = "https://media.istockphoto.com/vectors/no-image-available-sign-vector-id922962354?k=20&m=922962354&s=612x612&w=0&h=f-9tPXlFXtz9vg_-WonCXKCdBuPUevOBkp3DQ-i0xqo=";
                        }}
                        src={p.thumbnail}
                        height={250}
                        width={250}
                        style={{ padding: "0.3rem", cursor: "pointer" }}
                        onClick={() => handleViewProduct(p.id)}
                      />
                      <Card.Body>
                        <div style={{ float: "right", marginLeft: "4px" }}>
                          <input
                            type="checkbox"
                            name="delete"
                            onChange={() => selectMultipleProduct(p.id)}
                          />
                        </div>
                        {customField.name &&
                          <Card.Title className='card-text-formatter'>
                            {p.name}
                          </Card.Title>}
                        {customField.manufacture &&
                          <Card.Title className='card-text-formatter'>
                            {p.manufacture}
                          </Card.Title>}
                        {customField.rating &&
                          <Card.Title className='card-text-formatter'>
                            <ReactStars
                              edit={false}
                              size={25}
                              value={parseInt(p.rating)}
                            />
                          </Card.Title>}
                        {customField.description &&
                          <Card.Text className='card-text-formatter'>
                            {p.description}
                          </Card.Text>
                        }
                        <div className='button-wrapper'>
                          <Button variant="primary" onClick={() => handleEditProduct(p.id)}>Edit <AiOutlineEdit style={{ marginTop: "-2px" }} /></Button>
                          <Button variant="danger" onClick={() => handleShow(p.id)}>Delete <RiDeleteBin6Line style={{ marginTop: "-2px" }} /></Button>
                        </div>
                      </Card.Body>
                    </Card>
                  </Col>
                )
              }) : <h5>No Product Found</h5>}
            </Row>
          </Container >
      }
    </>
  )
}

export default ProductList