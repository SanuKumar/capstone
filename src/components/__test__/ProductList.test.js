import ProductList from '../ProductList';
import { cleanup, screen, render } from "@testing-library/react"

afterEach(cleanup)

test('renders ProductList correctly when there are empy products array', () => {
  const products = []
  const { asFragment } = render(<ProductList products={products} />)
  expect(asFragment()).toMatchSnapshot();
});

test("Product Field Header", () => {
  render(<ProductList />)
  const header = screen.getByTestId('product-field-header')
  expect(header.textContent).toBe('Customise Product Fields')
})

describe("When products array props passed to ProductList component is null", () => {
  render(<ProductList products={null} />)

  test("should not crash and no Product Component is rendered", () => {
    render(<ProductList />)
  })
})


test("Check if Add button is rendered", () => {
  render(<ProductList />)
  expect(screen.getByRole('button', { name: /add product/i })).toHaveTextContent('Add Product')
})
