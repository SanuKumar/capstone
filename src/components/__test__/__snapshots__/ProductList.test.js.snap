// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProductList correctly when there are empy products array 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container-wrapper container"
  >
    <div
      class="row"
    >
      <div
        class="col-lg-10 col-md-10 col-sm-12"
      >
        <h2
          data-testid="product-field-header"
        >
          Customise Product Fields
          <div
            style="height: 2px; width: 2px;"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                class="dropdown-toggle btn btn-warning"
                id="dropdown-basic-button"
                type="button"
              >
                <svg
                  enable-background="new 0 0 48 48"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  version="1"
                  viewBox="0 0 48 48"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    fill="#F57C00"
                    points="29,23 19,23 7,9 41,9"
                  />
                  <g
                    fill="#FF9800"
                  >
                    <polygon
                      points="29,38 19,44 19,23 29,23"
                    />
                    <path
                      d="M41.5,9h-35C5.7,9,5,8.3,5,7.5v0C5,6.7,5.7,6,6.5,6h35C42.3,6,43,6.7,43,7.5v0C43,8.3,42.3,9,41.5,9z"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
        </h2>
        <form
          class="form-wrapper"
        >
          <label>
            Product Name
          </label>
          <input
            checked=""
            name="name"
            type="checkbox"
          />
          <label>
            Manufacture
          </label>
          <input
            checked=""
            name="manufacture"
            type="checkbox"
          />
          <label>
            Rating
          </label>
          <input
            checked=""
            name="rating"
            type="checkbox"
          />
          <label>
            Description
          </label>
          <input
            checked=""
            name="description"
            type="checkbox"
          />
        </form>
      </div>
      <div
        class="col"
      >
        <div
          class="row"
        >
          <div
            class="m-1 col"
          >
            <div
              style="margin: 2px; width: 10.5rem;"
            />
          </div>
          <div
            class="m-1 col"
          >
            <div
              style="margin: 2px; width: 9rem;"
            >
              <button
                class="btn btn-primary"
                data-testid="add-product-btn"
                type="button"
              >
                Add Product 
                <svg
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="margin-top: -2px;"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 6C12.5523 6 13 6.44772 13 7V11H17C17.5523 11 18 11.4477 18 12C18 12.5523 17.5523 13 17 13H13V17C13 17.5523 12.5523 18 12 18C11.4477 18 11 17.5523 11 17V13H7C6.44772 13 6 12.5523 6 12C6 11.4477 6.44772 11 7 11H11V7C11 6.44772 11.4477 6 12 6Z"
                    fill="currentColor"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M5 22C3.34315 22 2 20.6569 2 19V5C2 3.34315 3.34315 2 5 2H19C20.6569 2 22 3.34315 22 5V19C22 20.6569 20.6569 22 19 22H5ZM4 19C4 19.5523 4.44772 20 5 20H19C19.5523 20 20 19.5523 20 19V5C20 4.44772 19.5523 4 19 4H5C4.44772 4 4 4.44772 4 5V19Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div
    class="container"
  >
    <div
      class="justify-content-md-center row"
    >
      <h5>
        No Product Found
      </h5>
    </div>
  </div>
</DocumentFragment>
`;
